### PR TITLE
Fix a regression in the discrete PC Speaker affecting Narco Police

### LINKED ISSUE #1776: PC speaker emulation causes Narco Police to segfault
1. Start Narco Police
2. Keep moving forward by pressing 9
3. When the first wave of enemy soldiers appear, move fully to the side by pressing 6 and start shooting with 0
4. 9 times out of 10 the game will segfault within seconds (interestingly it happens much less frequently when you stay in the centre)

If you disable the PC speaker emulation, the game doesn't crash.

Sanitizer logs:

```
2022-07-07 22:59:23.256 | ENVELOPE: PCSPEAKER done after 7142 frames, peak sample was 16399.000000
2022-07-07 22:59:35.678 | EXCEPTION: Standard library exception: array::at
2022-07-07 22:59:35.790 | atexit
2022-07-07 22:59:35.791 | PCSPEAKER: Shutting down discrete model
/opt/local/libexec/llvm-14/bin/../include/c++/v1/__memory/unique_ptr.h:503:12: runtime error: applying non-zero offset 800 to null pointer
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /opt/local/libexec/llvm-14/bin/../include/c++/v1/__memory/unique_ptr.h:503:12 in

Loguru caught a signal: SIGSEGV
2022-07-07 22:59:35.792 | Signal: SIGSEGV
zsh: segmentation fault  ~/dev/jn/dosbox-staging/build/sanitizer/dosbox
```

### PR DESCRIPTION
**Pull request by kcgen**
Fixes #1776

@johnnovak, thanks for catching this.

### CODE REVIEW

File: src/hardware/pcspeaker_discrete.cpp [hunk 1]
```diff
@@ -341,7 +338,7 @@ void PcSpeakerDiscrete::SetType(const PpiPortB &b)
 }
 
 // Halt the channel after the speaker has idled
-void PcSpeakerDiscrete::PlayOrSleep(const uint16_t speaker_movements,
+void PcSpeakerDiscrete::PlayOrSleep(const bool speaker_movements,
```
Comment (johnnovak) [inline src/hardware/pcspeaker_discrete.cpp:341]: Just by looking at the name, I only have a very vague idea about what "speaker_movements" signifies. Could you find a better name perhaps?

File: src/hardware/pcspeaker_discrete.cpp [hunk 2]
```diff
@@ -362,12 +359,12 @@ void PcSpeakerDiscrete::ChannelCallback(const uint16_t frames)
...
-	uint16_t pos          = 0u;
-	auto sample_base      = 0.0f;
+	last_index       = 0.0f;
+	auto sample_base = 0.0f;
+
+	bool enteries_were_dequeued = false;
 
-	const auto period_per_frame_ms = 1.0f / frames - FLT_EPSILON;
-	// The epsilon reduction ensures the sum of the periods is <= 1ms
+	const auto period_per_frame_ms = 1.0001f / frames;
```
Comment (johnnovak) [inline src/hardware/pcspeaker_discrete.cpp:367]: Why not just 1.0f?

File: src/hardware/pcspeaker_discrete.h [hunk 3]
```diff
@@ -67,7 +67,7 @@ class PcSpeakerDiscrete final : public PcSpeaker {
 		float vol   = 0.0f;
 	};
 
-	std::array<DelayEntry, 1024> entries = {};
+	std::queue<DelayEntry> entries = {};
```
Comment (johnnovak) [inline src/hardware/pcspeaker_discrete.h:70]: 👍🏻

**Review by johnnovak — APPROVED** [review]
You fixed this pretty quickly @kcgen, great work! 👍🏻

File: src/hardware/pcspeaker_discrete.cpp (hunk 1 above)
Comment (kcgen) [reply src/hardware/pcspeaker_discrete.cpp]: Yes - now "samples_were_processed".

The good news is this all disappears when we move to auto-sleep.

File: src/hardware/pcspeaker_discrete.h (hunk 3 above)
Comment (kcgen) [reply src/hardware/pcspeaker_discrete.h:70]: Indeed; now repeated gun fire won't take us down!

File: src/hardware/pcspeaker_discrete.cpp (hunk 2 above)
Comment (kcgen) [reply src/hardware/pcspeaker_discrete.cpp]: Deja vu: https://github.com/dosbox-staging/dosbox-staging/pull/1690#discussion_r902189308

This factor is now better understood, and I've described it as a comment in the code.

**Comment by kcgen** [comment]
> You fixed this pretty quickly @kcgen, great work! 👍🏻

("quickly" - but you don't want to see the hours it took :sweat_smile: )

**Comment by kcgen** [comment]
Thanks for the quick review and testing, @johnnovak!

**Pull request merged**